### PR TITLE
feat(utility): implement "LazyList.__setitem__"

### PR DESCRIPTION
**Pull request by linjiX**

### CODE REVIEW

**Comment by GravitiRob** [comment]
Task linked: [CU-2846h64 实现 LazyList.__setitem__ 方法](https://app.clickup.com/t/2846h64)

File: graviti/utility/lazy2.py [hunk 1]
```diff
@@ -117,23 +200,59 @@ class LazyPage:
...
         return self._get_page().__iter__()  # type: ignore[no-any-return]
 
 
+class LazySlicedPage(LazyPage):
+    """LazyPage is a placeholder of the pages in the lazy list when the page is not loaded yet.
+
+    Arguments:
+        pos: The page number.
+        parent: The parent lazy list.
+
```
Comment (Lee-000) [inline graviti/utility/lazy2.py:242]: The docstring needs to be updated. Are the arguments necessary?

File: graviti/utility/lazy2.py [hunk 2]
```diff
@@ -117,23 +200,59 @@ class LazyPage:
 
     """
 
-    def __init__(self, pos: int, parent: LazyList) -> None:
-        self._pos = pos
+    def __init__(self, pos: int, ranging: range, parent: LazyList) -> None:
```
Comment (Lee-000) [inline graviti/utility/lazy2.py:203]: The docstring needs to be updated.

File: graviti/utility/lazy2.py [hunk 3]
```diff
@@ -28,8 +28,35 @@ def __init__(self, total_count: int, limit: int) -> None:
...
+        except IndexError:
+            last_offset = self.total_count
+
+        diff = partial_offsets.pop() - last_offset
+        if diff != 0:
+            self.total_count += diff
+            for i in range(start + 1, len(offsets)):
+                offsets[i] += diff
+
+        offsets[start : stop + 1] = partial_offsets
```
Comment (Lee-000) [inline graviti/utility/lazy2.py:59]: Why not update the offsets after setting the partial_offsets?

File: graviti/utility/lazy2.py [hunk 4]
```diff
@@ -96,6 +144,41 @@ def _getitem(self, index: int) -> Any:
...
+        right_page = self._pages[stop_i][stop_j + 1 :]
+        right_length = len(right_page)
+        if right_length:
+            pages.append(right_page)
+            lengths.append(right_length)
+
+        self._pages[start_i : stop_i + 1] = pages
+        self._offsets.setitem(start_i, stop_i, lengths)
+
+        self._update_parent_pos(start_i, stop_i - start_i + len(pages) - 1)
```
Comment (Lee-000) [inline graviti/utility/lazy2.py:172]: Why start from `start_i`? Should it start from `start_i + len(pages)`?
And why the pos_offset is calculated like this?

File: graviti/utility/lazy2.py (hunk 3 above)
Comment (linjiX) [reply graviti/utility/lazy2.py:59]: because if the diff is not 0, the offsets needs to be updated first.

File: graviti/utility/lazy2.py [hunk 5]
```diff
@@ -88,6 +113,29 @@ def __len__(self) -> int:
...
+        ...
+
+    @overload
+    def __setitem__(self, index: slice, value: Iterable[Any]) -> None:
+        ...
+
+    def __setitem__(self, index: Union[int, slice], value: Union[Any, Iterable[Any]]) -> None:
+        if isinstance(index, slice):
+            start, stop, step = index.indices(self.__len__())
+            assert step == 1
```
Comment (AChenQ) [inline graviti/utility/lazy2.py:127]: assert ?

**Review by AChenQ — CHANGES_REQUESTED** [review]

File: graviti/utility/lazy2.py (hunk 4 above)
Comment (linjiX) [reply graviti/utility/lazy2.py:172]: Here is a bug, it should not start from `start_i`, it should start from `start_i + 1`, the left_page also needs to be updated, so it should not start from `start_i + len(pages)`

Comment (linjiX) [reply graviti/utility/lazy2.py:172]:
```python
L1 = [...]
L2 = [...]
L1[start:stop] = L2
```
The items in L1[start + 1:] will shift `stop - start + len(L2) - 1`

File: graviti/utility/lazy2.py (hunk 5 above)
Comment (linjiX) [reply graviti/utility/lazy2.py:127]: the step == 1 case is not supported yet

**Review by AChenQ — APPROVED** [review]

**Review by Lee-000 — APPROVED** [review]

**Pull request merged**